src/LtiServiceConnector.php

```diff
@@ -18,39 +19,36 @@ class LtiServiceConnector implements ILtiServiceConnector
 
     private $cache;
     private $client;
-    private $registration;
-    private $access_tokens = [];
 
-    public function __construct(ILtiRegistration $registration, ICache $cache, Client $client)
+    public function __construct(ICache $cache, Client $client)
     {
-        $this->registration = $registration;
         $this->cache = $cache;
         $this->client = $client;
     }
 
-    public function getAccessToken(array $scopes)
+    public function getAccessToken(ILtiRegistration $registration, array $scopes)
```

Comment:
hmm what's the idea behind moving registration out of the constructor? It seems like it adds more overhead to have to now pass it into individual functions?

Reply:
**TLDR:** LTI Services (like AGS, NRPS) _should_ be tied to a specific LMS (LtiRegistration), but there's no reason the underlying client they use needs to be.

The `LtiRegistration` basically represents a relationship to a single LMS integration. By having it in the constructor, it means that when you instantiate an LtiServiceConnector, it becomes inextricably tied to that LMS integration. If you want to make service calls to a different LMS, you have to create a completely new instance of the service connector.

The service connector basically acts like an API Client, dispatching calls on behalf of various LTI Services. It's an abstraction to handle that handles making API requests and returning data. I couldn't come up with a good reason to couple each instance to a specific LtiRegistration. It seemed better to build it such that the Service Connector doesn't have to be concerned about which LMS it belongs to or which LTI Service is using it.

In contrast, the AGS and NRPS are tied to specific LtiRegistrations. When we're syncing grades, those _are_ specific to a single LMS integration (LtiRegistration). It made sense to me to say "this instance of AGS can only sync grades to this LMS". Hence, I moved the LtiRegistration to the constructor of the various LTI Services.

Reply:
Ohhh I see that makes sense, sorry you had to explain that again. It finally clicked for me. Thanks Davo!